File: core/harness-adapter/src/test/kotlin/com/deepseek/harness/android/harness/TimelineReducerTest.kt
package com.deepseek.harness.android.harness

import com.deepseek.harness.android.domain.model.JobStatus
import com.deepseek.harness.android.domain.model.MessageRole
import com.deepseek.harness.android.domain.model.QuestionIntent
import com.deepseek.harness.android.domain.model.TimelineItem
import com.deepseek.harness.android.domain.model.ToolRunStatus
import com.deepseek.harness.android.network.ServerRequest
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class TimelineReducerTest {

    @Test
    fun `history finalizes live assistant text from chunks`() {
        val history = listOf(
            event(
                seq = 1L,
                type = "assistant/chunk",
                data = buildJsonObject {
                    put("turn", 1)
                    put("step", 1)
                    put("chunk", buildJsonObject { put("type", "text-delta"); put("index", 0); put("text", "hello") })
                },
            ),
            event(
                seq = 2L,
                type = "assistant/message",
                data = buildJsonObject {
                    put("turn", 1)
                    put("step", 1)
                    put(
                        "message",
                        buildJsonObject {
                            put("id", "assistant-1")
                            put("role", "assistant")
                            put("content", buildJsonArray { add(textBlock("hello")) })
                        },
                    )
                },
            ),
        )

        val reducer = TimelineReducer("s1")
        reducer.reset(history)

        val snapshot = reducer.snapshot()
        assertEquals(1, snapshot.size)
        val message = snapshot.single() as TimelineItem.Message
        assertEquals("assistant-1", message.value.id)
        assertEquals(MessageRole.ASSISTANT, message.value.role)
        assertEquals("hello", message.value.text)
        assertEquals(false, message.value.streaming)
    }

    @Test
    fun `tool call pairs with result`() {
        val history = listOf(
            event(
                seq = 1L,
                type = "tool/call",
                data = buildJsonObject {
                    put("turn", 1)
                    put("step", 1)
                    put("callId", "call-1")
                    put("name", "bash")
                    put("arguments", "{}")
                },
            ),
            event(
                seq = 2L,
                type = "tool/result",
                data = buildJsonObject {
                    put("turn", 1)
                    put("step", 1)
                    put(
                        "message",
                        buildJsonObject {
                            put(
                                "content",
                                buildJsonArray {
                                    add(
                                        buildJsonObject {
                                            put("type", "tool-result")
                                            put("toolCallId", "call-1")
                                            put(
                                                "content",
                                                buildJsonArray { add(textBlock("ok")) },
                                            )
                                        },
                                    )
                                },
                            )
                        },
                    )
                },
            ),
        )

        val reducer = TimelineReducer("s1")
        reducer.reset(history)

        val tool = reducer.snapshot().single() as TimelineItem.ToolCall
        assertEquals("call-1", tool.id)
        assertEquals("bash", tool.name)
        assertEquals(ToolRunStatus.COMPLETED, tool.status)
        assertEquals("ok", tool.result)
    }

    @Test
    fun `tool result block isError marks the paired call failed`() {
        val history = listOf(
            event(
                seq = 1L,
                type = "tool/call",
                data = buildJsonObject {
                    put("turn", 1)
                    put("step", 1)
                    put("callId", "call-failed")
                    put("name", "bash")
                    put("arguments", "{}")
                },
            ),
            event(
                seq = 2L,
                type = "tool/result",
                data = buildJsonObject {
                    put("turn", 1)
                    put("step", 1)
                    put(
                        "message",
                        buildJsonObject {
                            put(
                                "content",
                                buildJsonArray {
                                    add(
                                        buildJsonObject {
                                            put("type", "tool-result")
                                            put("toolCallId", "call-failed")
                                            put(
                                                "content",
                                                buildJsonArray { add(textBlock("boom")) },
                                            )
                                            put("isError", true)
                                        },
                                    )
                                },
                            )
                        },
                    )
                },
            ),
        )

        val reducer = TimelineReducer("s1")
        reducer.reset(history)

        val tool = reducer.snapshot().single() as TimelineItem.ToolCall
        assertEquals(ToolRunStatus.FAILED, tool.status)
        assertEquals(true, tool.isError)
        assertEquals("boom", tool.result)
    }

    @Test
    fun `approval frame becomes answerable card and resolved removes it`() {
        val reducer = TimelineReducer("s1")
        val requested = ServerRequest(
            type = "server-request",
            rpcId = "rpc-1",
            method = "approval/requested",
            payload = buildJsonObject {
                put("type", "approval/requested")
                put("sessionId", "s1")
                put("approvalId", "approval-1")
                put("toolName", "bash")
                put("reason", "run command")
            },
        )
        val resolved = ServerRequest(
            type = "server-request",
            rpcId = "rpc-2",
            method = "approval/resolved",
            payload = buildJsonObject {
                put("type", "approval/resolved")
                put("sessionId", "s1")
                put("approvalId", "approval-1")
                put("outcome", "allowed-once")
            },
        )

        reducer.ingestFrame(requested)
        val approval = reducer.snapshot().single() as TimelineItem.ApprovalRequest
        assertEquals("rpc-1", approval.requestId)
        assertEquals("bash", approval.toolName)

        reducer.ingestFrame(resolved)
        assertTrue(reducer.snapshot().isEmpty())
    }

    @Test
    fun `question intent parses for plan review presentation`() {
        val reducer = TimelineReducer("s1")
        reducer.ingestFrame(
            ServerRequest(
                type = "server-request",
                rpcId = "rpc-plan",
                method = "question/requested",
                payload = buildJsonObject {
                    put("type", "question/requested")
                    put("sessionId", "s1")
                    put(
                        "questions",
                        buildJsonArray {
                            add(
                                buildJsonObject {
                                    put("id", "plan-review")
                                    put("header", "Plan review")
                                    put("question", "Approve this plan and leave plan mode?")
                                    put("detail", "# The plan")
                                    put(
                                        "options",
                                        buildJsonArray {
                                            add(buildJsonObject { put("label", "Approve plan") })
                                            add(buildJsonObject { put("label", "Keep planning") })
                                        },
                                    )
                                    put(
                                        "intent",
                                        buildJsonObject {
                                            put("kind", "plan-review")
                                            put("approve", "Approve plan")
                                        },
                                    )
                                },
                            )
                        },
                    )
                },
            ),
        )

        val question = (reducer.snapshot().single() as TimelineItem.QuestionRequest).questions.single()
        assertEquals(QuestionIntent(kind = "plan-review", approve = "Approve plan"), question.intent)
        assertEquals("# The plan", question.detail)
    }

    @Test
    fun `question frame keeps header and option descriptions and resolution removes it`() {
        val reducer = TimelineReducer("s1")
        val requested = ServerRequest(
            type = "server-request",
            rpcId = "rpc-question",
            method = "question/requested",
            payload = buildJsonObject {
                put("type", "question/requested")
                put("sessionId", "s1")
                put(
                    "questions",
                    buildJsonArray {
                        add(
                            buildJsonObject {
                                put("id", "q1")
                                put("header", "Before continuing")
                                put("question", "Proceed?")
                                put("detail", "This command writes files")
                                put("multiSelect", true)
                                put(
                                    "options",
                                    buildJsonArray {
                                        add(
                                            buildJsonObject {
                                                put("label", "yes")
                                                put("description", "Continue now")
                                            },
                                        )
                                        add(buildJsonObject { put("label", "no") })
                                    },
                                )
                            },
                        )
                    },
                )
            },
        )

        reducer.ingestFrame(requested)

        val question = reducer.snapshot().single() as TimelineItem.QuestionRequest
        val item = question.questions.single()
        assertEquals("q1", item.id)
        assertEquals("Before continuing", item.header)
        assertTrue(item.multiSelect)
        assertEquals("Continue now", item.optionDescriptions["yes"])
        assertEquals(listOf("yes", "no"), item.options)

        reducer.ingestFrame(
            ServerRequest(
                type = "server-request",
                rpcId = "rpc-resolved",
                method = "question/resolved",
                payload = buildJsonObject {
                    put("type", "question/resolved")
                    put("sessionId", "s1")
                    put("questionRpcId", "rpc-question")
                    put("outcome", "answered")
                },
            ),
        )
        assertTrue(reducer.snapshot().isEmpty())
    }

    @Test
    fun `session jobs frame becomes job snapshot`() {
        val reducer = TimelineReducer("s1")
        val frame = ServerRequest(
            type = "server-request",
            rpcId = "rpc-jobs",
            method = "session/jobs",
            payload = buildJsonObject {
                put("type", "session/jobs")
                put("sessionId", "s1")
                put(
                    "jobs",
                    buildJsonArray {
                        add(
                            buildJsonObject {
                                put("id", "bash-1")
                                put("kind", "bash")
                                put("label", "sleep 10")
                                put("status", "running")
                                put("startedAt", 10L)
                            },
                        )
                    },
                )
            },
        )

        reducer.ingestFrame(frame)

        val jobs = reducer.snapshot().single() as TimelineItem.Jobs
        assertEquals(1, jobs.jobs.size)
        assertEquals("bash-1", jobs.jobs.single().id)
        assertEquals(JobStatus.RUNNING, jobs.jobs.single().status)
    }

    @Test
    fun `user message image blocks fold into attachment refs`() {
        val history = listOf(
            event(
                seq = 1L,
                type = "user/message",
                data = buildJsonObject {
                    put("id", "user-1")
                    put(
                        "content",
                        buildJsonArray {
                            add(textBlock("look at this"))
                            add(
                                buildJsonObject {
                                    put("type", "image")
                                    put(
                                        "attachment",
                                        buildJsonObject {
                                            put("attachmentId", "sha256:abc")
                                            put("mediaType", "image/png")
                                            put("bytes", 2048)
                                            put("width", 640)
                                            put("height", 480)
                                            put("name", "shot.png")
                                        },
                                    )
                                },
                            )
                        },
                    )
                },
            ),
        )

        val reducer = TimelineReducer("s1")
        reducer.reset(history)

        val message = (reducer.snapshot().single() as TimelineItem.Message).value
        assertEquals("look at this", message.text)
        val image = message.images.single()
        assertEquals("sha256:abc", image.attachmentId)
        assertEquals("image/png", image.mediaType)
        assertEquals(2048L, image.bytes)
        assertEquals(640, image.width)
        assertEquals(480, image.height)
        assertEquals("shot.png", image.name)
    }

    @Test
    fun `turn start events become boundaries and dedupe per turn`() {
        val history = listOf(
            event(
                seq = 1L,
                type = "turn/start",
                data = buildJsonObject { put("turn", 1) },
            ),
            event(
                seq = 2L,
                type = "user/message",
                data = buildJsonObject { put("id", "user-1"); put("content", buildJsonArray { add(textBlock("hi")) }) },
            ),
            event(
                seq = 3L,
                type = "turn/start",
                data = buildJsonObject { put("turn", 1) },
            ),
            event(
                seq = 4L,
                type = "turn/start",
                data = buildJsonObject { put("turn", 2) },
            ),
            event(
                seq = 5L,
                type = "user/message",
                data = buildJsonObject { put("id", "user-2"); put("content", buildJsonArray { add(textBlock("again")) }) },
            ),
        )

        val reducer = TimelineReducer("s1")
        reducer.reset(history)

        val snapshot = reducer.snapshot()
        val turns = snapshot.filterIsInstance<TimelineItem.TurnBoundary>()
        assertEquals(listOf(1L, 2L), turns.map { it.turn })
        assertEquals(0, snapshot.indexOfFirst { it is TimelineItem.TurnBoundary && it.turn == 1L })
        assertEquals(2, snapshot.indexOfFirst { it is TimelineItem.TurnBoundary && it.turn == 2L })
    }

    private fun event(seq: Long, type: String, data: JsonObject): JsonObject =
        buildJsonObject {
            put("type", type)
            put("seq", seq)
            put("time", 1L)
            put("data", data)
        }

    private fun textBlock(text: String): JsonObject =
        buildJsonObject {
            put("type", "text")
            put("text", text)
        }
}

File: flutter/packages/harness_adapter/test/timeline_reducer_test.dart
import 'package:domain/model/jobs.dart';
import 'package:domain/model/timeline_item.dart';
import 'package:domain/model/chat_message.dart';
import 'package:network/rpc_envelope.dart';
import 'package:test/test.dart';

import 'package:harness_adapter/src/rpc_map.dart';
import 'package:harness_adapter/src/timeline_reducer.dart';

JsonMap event(int seq, String type, JsonMap data) => <String, Object?>{
      'type': type,
      'seq': seq,
      'time': 1,
      'data': data,
    };

JsonMap textBlock(String text) => <String, Object?>{
      'type': 'text',
      'text': text,
    };

void main() {
  test('history finalizes live assistant text from chunks', () {
    final history = <JsonMap>[
      event(
        1,
        'assistant/chunk',
        <String, Object?>{
          'turn': 1,
          'step': 1,
          'chunk': <String, Object?>{
            'type': 'text-delta',
            'index': 0,
            'text': 'hello',
          },
        },
      ),
      event(
        2,
        'assistant/message',
        <String, Object?>{
          'turn': 1,
          'step': 1,
          'message': <String, Object?>{
            'id': 'assistant-1',
            'role': 'assistant',
            'content': <Object?>[textBlock('hello')],
          },
        },
      ),
    ];

    final reducer = TimelineReducer('s1');
    reducer.reset(history);

    final snapshot = reducer.snapshot();
    expect(snapshot, hasLength(1));
    final message = snapshot.single as TimelineMessage;
    expect(message.value.id, 'assistant-1');
    expect(message.value.role, MessageRole.assistant);
    expect(message.value.text, 'hello');
    expect(message.value.streaming, false);
  });

  test('tool call pairs with result', () {
    final history = <JsonMap>[
      event(
        1,
        'tool/call',
        <String, Object?>{
          'turn': 1,
          'step': 1,
          'callId': 'call-1',
          'name': 'bash',
          'arguments': '{}',
        },
      ),
      event(
        2,
        'tool/result',
        <String, Object?>{
          'turn': 1,
          'step': 1,
          'message': <String, Object?>{
            'content': <Object?>[
              <String, Object?>{
                'type': 'tool-result',
                'toolCallId': 'call-1',
                'content': <Object?>[textBlock('ok')],
              },
            ],
          },
        },
      ),
    ];

    final reducer = TimelineReducer('s1');
    reducer.reset(history);

    final tool = reducer.snapshot().single as TimelineToolCall;
    expect(tool.id, 'call-1');
    expect(tool.name, 'bash');
    expect(tool.status, ToolRunStatus.completed);
    expect(tool.result, 'ok');
  });

  test('tool result block isError marks the paired call failed', () {
    final history = <JsonMap>[
      event(
        1,
        'tool/call',
        <String, Object?>{
          'turn': 1,
          'step': 1,
          'callId': 'call-failed',
          'name': 'bash',
          'arguments': '{}',
        },
      ),
      event(
        2,
        'tool/result',
        <String, Object?>{
          'turn': 1,
          'step': 1,
          'message': <String, Object?>{
            'content': <Object?>[
              <String, Object?>{
                'type': 'tool-result',
                'toolCallId': 'call-failed',
                'content': <Object?>[textBlock('boom')],
                'isError': true,
              },
            ],
          },
        },
      ),
    ];

    final reducer = TimelineReducer('s1');
    reducer.reset(history);

    final tool = reducer.snapshot().single as TimelineToolCall;
    expect(tool.status, ToolRunStatus.failed);
    expect(tool.isError, true);
    expect(tool.result, 'boom');
  });

  test('approval frame becomes answerable card and resolved removes it', () {
    final reducer = TimelineReducer('s1');
    final requested = ServerRequest(
      rpcId: 'rpc-1',
      method: 'approval/requested',
      payload: <String, Object?>{
        'type': 'approval/requested',
        'sessionId': 's1',
        'approvalId': 'approval-1',
        'toolName': 'bash',
        'reason': 'run command',
      },
    );
    final resolved = ServerRequest(
      rpcId: 'rpc-2',
      method: 'approval/resolved',
      payload: <String, Object?>{
        'type': 'approval/resolved',
        'sessionId': 's1',
        'approvalId': 'approval-1',
        'outcome': 'allowed-once',
      },
    );

    reducer.ingestFrame(requested);
    final approval = reducer.snapshot().single as TimelineApprovalRequest;
    expect(approval.requestId, 'rpc-1');
    expect(approval.toolName, 'bash');

    reducer.ingestFrame(resolved);
    expect(reducer.snapshot(), isEmpty);
  });

  test('question intent parses for plan review presentation', () {
    final reducer = TimelineReducer('s1');
    reducer.ingestFrame(
      ServerRequest(
        rpcId: 'rpc-plan',
        method: 'question/requested',
        payload: <String, Object?>{
          'type': 'question/requested',
          'sessionId': 's1',
          'questions': <Object?>[
            <String, Object?>{
              'id': 'plan-review',
              'header': 'Plan review',
              'question': 'Approve this plan and leave plan mode?',
              'detail': '# The plan',
              'options': <Object?>[
                <String, Object?>{'label': 'Approve plan'},
                <String, Object?>{'label': 'Keep planning'},
              ],
              'intent': <String, Object?>{
                'kind': 'plan-review',
                'approve': 'Approve plan',
              },
            },
          ],
        },
      ),
    );

    final question =
        (reducer.snapshot().single as TimelineQuestionRequest).questions.single;
    expect(question.intent,
        const QuestionIntent(kind: 'plan-review', approve: 'Approve plan'));
    expect(question.detail, '# The plan');
  });

  test('question frame keeps header and option descriptions and resolution removes it', () {
    final reducer = TimelineReducer('s1');
    final requested = ServerRequest(
      rpcId: 'rpc-question',
      method: 'question/requested',
      payload: <String, Object?>{
        'type': 'question/requested',
        'sessionId': 's1',
        'questions': <Object?>[
          <String, Object?>{
            'id': 'q1',
            'header': 'Before continuing',
            'question': 'Proceed?',
            'detail': 'This command writes files',
            'multiSelect': true,
            'options': <Object?>[
              <String, Object?>{
                'label': 'yes',
                'description': 'Continue now',
              },
              <String, Object?>{'label': 'no'},
            ],
          },
        ],
      },
    );

    reducer.ingestFrame(requested);

    final question = reducer.snapshot().single as TimelineQuestionRequest;
    final item = question.questions.single;
    expect(item.id, 'q1');
    expect(item.header, 'Before continuing');
    expect(item.multiSelect, isTrue);
    expect(item.optionDescriptions['yes'], 'Continue now');
    expect(item.options, <String>['yes', 'no']);

    reducer.ingestFrame(
      ServerRequest(
        rpcId: 'rpc-resolved',
        method: 'question/resolved',
        payload: <String, Object?>{
          'type': 'question/resolved',
          'sessionId': 's1',
          'questionRpcId': 'rpc-question',
          'outcome': 'answered',
        },
      ),
    );
    expect(reducer.snapshot(), isEmpty);
  });

  test('session jobs frame becomes job snapshot', () {
    final reducer = TimelineReducer('s1');
    final frame = ServerRequest(
      rpcId: 'rpc-jobs',
      method: 'session/jobs',
      payload: <String, Object?>{
        'type': 'session/jobs',
        'sessionId': 's1',
        'jobs': <Object?>[
          <String, Object?>{
            'id': 'bash-1',
            'kind': 'bash',
            'label': 'sleep 10',
            'status': 'running',
            'startedAt': 10,
          },
        ],
      },
    );

    reducer.ingestFrame(frame);

    final jobs = reducer.snapshot().single as TimelineJobs;
    expect(jobs.jobs, hasLength(1));
    expect(jobs.jobs.single.id, 'bash-1');
    expect(jobs.jobs.single.status, JobStatus.running);
  });

  test('user message image blocks fold into attachment refs', () {
    final history = <JsonMap>[
      event(
        1,
        'user/message',
        <String, Object?>{
          'id': 'user-1',
          'content': <Object?>[
            textBlock('look at this'),
            <String, Object?>{
              'type': 'image',
              'attachment': <String, Object?>{
                'attachmentId': 'sha256:abc',
                'mediaType': 'image/png',
                'bytes': 2048,
                'width': 640,
                'height': 480,
                'name': 'shot.png',
              },
            },
          ],
        },
      ),
    ];

    final reducer = TimelineReducer('s1');
    reducer.reset(history);

    final message = (reducer.snapshot().single as TimelineMessage).value;
    expect(message.text, 'look at this');
    final image = message.images.single;
    expect(image.attachmentId, 'sha256:abc');
    expect(image.mediaType, 'image/png');
    expect(image.bytes, 2048);
    expect(image.width, 640);
    expect(image.height, 480);
    expect(image.name, 'shot.png');
  });

  test('turn start events become boundaries and dedupe per turn', () {
    final history = <JsonMap>[
      event(1, 'turn/start', <String, Object?>{'turn': 1}),
      event(
        2,
        'user/message',
        <String, Object?>{
          'id': 'user-1',
          'content': <Object?>[textBlock('hi')],
        },
      ),
      event(3, 'turn/start', <String, Object?>{'turn': 1}),
      event(4, 'turn/start', <String, Object?>{'turn': 2}),
      event(
        5,
        'user/message',
        <String, Object?>{
          'id': 'user-2',
          'content': <Object?>[textBlock('again')],
        },
      ),
    ];

    final reducer = TimelineReducer('s1');
    reducer.reset(history);

    final snapshot = reducer.snapshot();
    final turns =
        snapshot.whereType<TimelineTurnBoundary>().toList();
    expect(turns.map((turn) => turn.turn), <int>[1, 2]);
    expect(
      snapshot.indexWhere((item) =>
          item is TimelineTurnBoundary && item.turn == 1),
      0,
    );
    expect(
      snapshot.indexWhere((item) =>
          item is TimelineTurnBoundary && item.turn == 2),
      2,
    );
  });

  test('compaction summary folds into a shadowed-count marker', () {
    final history = <JsonMap>[
      event(
        1,
        'compaction/summary',
        <String, Object?>{
          'compactionId': 'c-1',
          'shadowedSeqs': <Object?>[1, 2, 3, 4],
        },
      ),
    ];

    final reducer = TimelineReducer('s1');
    reducer.reset(history);

    final marker = reducer.snapshot().single as TimelineCompaction;
    expect(marker.shadowedCount, 4);
  });
}
